Library/Homebrew/cmd/update-report.rb

```diff
@@ -167,16 +170,113 @@ def migrate_legacy_cache_if_necessary
+    return unless ARGV.homebrew_developer?
+
+    unless HOMEBREW_PREFIX.writable_real?
+      onoe <<-EOS.undent
+        #{HOMEBREW_PREFIX} is not writable.
+
+        You should change the ownership and permissions of #{HOMEBREW_PREFIX}
+        temporarily back to your user account so we can complete the Homebrew
+        repository migration:
+          sudo chown -R $(whoami) #{HOMEBREW_PREFIX}
+      EOS
+    end
+
+    Keg::ALL_TOP_LEVEL_DIRECTORIES.each do |dir|
+      FileUtils.mkdir_p "#{HOMEBREW_PREFIX}/#{dir}"
+    end
+
+    new_homebrew_repository = Pathname.new "/usr/local/Homebrew"
+    if new_homebrew_repository.exist? && !new_homebrew_repository.children.empty?
+      onoe <<-EOS.undent
+        #{new_homebrew_repository} already exists.
+        Please remove it manually or uninstall and reinstall Homebrew into a new
+        location as the migration cannot be done automatically.
+      EOS
+    end
+    new_homebrew_repository.mkpath
+
+    repo_files = HOMEBREW_REPOSITORY.cd do
+      Utils.popen_read("git ls-files").lines.map(&:chomp)
+    end
+
+    FileUtils.cp_r "#{HOMEBREW_REPOSITORY}/.git", "#{new_homebrew_repository}/.git"
+    new_homebrew_repository.cd do
+      safe_system "git", "checkout", "."
+    end
+
+    if (HOMEBREW_REPOSITORY/"Library/Taps").exist?
+      FileUtils.cp_r "#{HOMEBREW_REPOSITORY}/Library/Taps", "#{new_homebrew_repository}/Library/Taps"
+    end
+
+    if (HOMEBREW_REPOSITORY/"Library/LinkedKegs").exist?
+      FileUtils.mkdir_p "#{new_homebrew_repository}/Library/LinkedKegs"
+      (HOMEBREW_REPOSITORY/"Library/LinkedKegs").children.each do |link|
+        src = link.realpath
+        dst = new_homebrew_repository/"Library/LinkedKegs/#{link.basename}"
+        FileUtils.ln_sf(dst.relative_path_from(src.parent), src)
+      end
+    end
+
+    src = Pathname.new("#{new_homebrew_repository}/bin/brew")
+    dst = Pathname.new("#{HOMEBREW_PREFIX}/bin/brew")
+    begin
+      FileUtils.ln_sf(dst.relative_path_from(src.parent), src)
+    rescue Errno::EACCES
+      onoe <<-EOS.undent
+        Could not create symlink at #{dst}!
+        Please do this manually with:
+          sudo ln -sf #{src} #{dst}
+          sudo chown $(whoami) #{dst}
+      EOS
+    end
+
+    unremovable_paths = []
+    remove_dirs = [".git", "share/doc/homebrew/images", "share/doc/homebrew"]
+    (repo_files + remove_dirs).each do |file|
+      path = Pathname.new "#{HOMEBREW_REPOSITORY}/#{file}"
+      begin
+        FileUtils.rm_rf path
+      rescue Errno::EACCES
+        unremovable_paths << path
+      end
+      path.parent.rmdir_if_possible
+    end
+
+    unless unremovable_paths.empty?
+      onoe <<-EOS.undent
+        Could not remove old HOMEBREW_REPOSITORY paths!
+        Please do this manually with:
+          sudo rm -rf #{unremovable_paths.join " "}
+      EOS
+    end
+
+    link_completions_and_docs(new_homebrew_repository)
+
+    ohai "Migrated HOMEBREW_REPOSITORY to #{new_homebrew_repository}!"
```

Comment:
Feels a little redundant to print `to /usr/local/Homebrew` twice in quick succession (the second time is the from ... to ... below). Maybe just use `HOMEBREW_REPOSITORY migrated!` as the title (or `Homebrew repository migrated!` since I `HOMEBREW_REPOSITORY` is an internal env var and users don't need to know about it).